pna.p4

```diff
@@ -545,6 +545,230 @@ extern Digest<T> {
 }
 // END:Digest_extern
 
+// BEGIN:ExactMap_extern
+struct exactmap_const_entry_t<K,V> {
+    K key;
+    V val;
+}
+
+struct exactmap_initial_entry_t<K,V> {
+    bool const_entry;
```

Comment:
Adding this bool "const" member to the struct complicates the already challenging semantics of const. It would be worth documenting a bit more clearly what it means. Does it mean it can't be deleted by any runtime API? What about automatic add/del ops? Does it change the table props in any other way (like current const does on tables).

Reply:
A short answer is that `const entries` in P4 tables means all entries are const, and none can be removed, modified, NOR added (this is in the language spec).  The constructor with `const_entries` is for these externs is intended to be like that.

The pull request that Andrew Pinski linked to is a proposed new thing in the language spec defining `entries` for a P4 table without a `const` qualifier.  See there for details, but the intent is that each of the entries can be independently marked as `const` or not, and only those marked `const` are prevented from being removed or modified by the control plane, but the control plane is free to add new entries to the table (up to the size limitation).  The constructor for these externs with `initial_entries` parameter is intended to work like that, and the true/false value for `const` takes the place of the `const` keyword.

While it would be nice if the same syntax could be used for P4 tables as here, that would be a bigger change to the spec, e.g. perhaps to define the current syntax in P4 tables as a syntax for a new type of object, perhaps called a 'table entry' object, but I wasn't hoping to go that far here.

Reply:
If any or all of the text in my previous comment could be usefully added to this PR, let me know.  I think some of it is said in comments in the pna.p4 include file, e.g. it says elsewere in this PR:

const_entries is a list of entries, similar to the 'const
     * entries' table property for tables.  This set of entries cannot
     * be removed or modified by the control plane, and also, the
     * control plane is not allowed to add any entries to an ExactMap
     * instance created using this constructor.

And for initial_entries it is commented as:

The same as the ExactMap constructor with a parameter named
     * `const_entries`, except that the control plane is allowed to
     * add entries to an ExactMap instance constructed using this
     * constructor (subject to capacity constraints, as usual), and
     * the control plane can modify or remove any entries that has a
     * `const_entry` field equal to false.  Any entries with a
     * `const_entry` field value equal to true cannot be modified or
     * removed by the control plane.